### ISSUE TITLE
Running circuit with qubits added directly fails

### ISSUE THREAD

**Issue by kevinsung** [Qiskit/qiskit-aer#1660]
### Informations

- **Qiskit Aer version**: 0.11.0
- **Python version**: 3.10.8
- **Operating system**: Arch Linux

### What is the current behavior?
```
Simulation failed and returned the following error message:
ERROR: Failed to load qobj: Invalid Qobj experiment: not enough memory slots.
Result(backend_name='aer_simulator', backend_version='0.11.0', qobj_id='', job_id='e612fa58-f665-4d7f-979e-00f59d727f5f', success=False, results=[], date=2022-11-17T07:50:39.243329, status=ERROR: Failed to load qobj: Invalid Qobj experiment: not enough memory slots., header=None, metadata={}, time_taken=0.00040602684020996094)
```

### Steps to reproduce the problem
```python
from qiskit import ClassicalRegister, QuantumCircuit, QuantumRegister
from qiskit.circuit import CircuitInstruction
from qiskit.circuit.library import Measure
from qiskit import Aer

qubits = QuantumRegister(2)
clbits = ClassicalRegister(2)

circuit = QuantumCircuit()
circuit.add_bits(qubits)
circuit.add_bits(clbits)
for q, c in zip(qubits, clbits):
    circuit.append(CircuitInstruction(Measure(), [q], [c]))

backend = Aer.get_backend("aer_simulator")
result = backend.run(circuit).result()
print(result)
```

### What is the expected behavior?
Should work. Terra can handle it.

**Comment by t-imamichi** [comment]
FYI: a circuit with `add_register` instead of `add_bits` works fine.

```python
from qiskit import ClassicalRegister, QuantumCircuit, QuantumRegister
from qiskit.circuit import CircuitInstruction
from qiskit.circuit.library import Measure
from qiskit import Aer

qubits = QuantumRegister(2)
clbits = ClassicalRegister(2)

circuit = QuantumCircuit()
circuit.add_register(qubits)
circuit.add_register(clbits)
for q, c in zip(qubits, clbits):
    circuit.append(CircuitInstruction(Measure(), [q], [c]))

backend = Aer.get_backend("aer_simulator")
result = backend.run(circuit).result()
print(result.get_counts())
```

output
```
{'00': 1024}
```

**Comment by jakelishman** [comment]
@t-imamichi: sure, but that's not the point of the bug being reported.  Loose bits should work.

This is almost certainly related to Aer's persistent use of the `Qobj` path for passing the objects down to C++, long past the point when that stopped making any sense, and `Qobj` simply cannot represent everything that needs to be represented.  Aer's Python/C++ boundary has been in need of rework for quite some time, and this is one more indication of that.

**Comment by kevinsung** [comment]
This bug affects usage of `QuantumCircuit.from_instructions` because that method uses `add_bits` instead of `add_register`.

**Comment by hhorii** [comment]
I guess that the current `assemble()` does not support classical bits added by `add_bits()`. In my understanding, `memory_slots` in QObj must be `2` but currently `0`.

```
from qiskit import ClassicalRegister, QuantumCircuit, QuantumRegister, assemble
from qiskit.circuit import CircuitInstruction
from qiskit.circuit.library import Measure
from qiskit import Aer

qubits = QuantumRegister(2)
clbits = ClassicalRegister(2)

circuit = QuantumCircuit()
circuit.add_bits(qubits)
circuit.add_bits(clbits)
# circuit.add_register(qubits)
# circuit.add_register(clbits)

print(assemble(circuit))
```

> QASM Qobj: 48d62624-1690-46a4-ba66-142b9504c8d8:
Config: {'init_qubits': True,
 'meas_level': <MeasLevel.CLASSIFIED: 2>,
 'memory': False,
 **'memory_slots': 0,**
 'n_qubits': 0,
 'parameter_binds': [],
 'parametric_pulses': [],
 'shots': 1024}
Header: {}
Experiments:
> 
> QASM Experiment:
Header:
{'clbit_labels': [],
 'creg_sizes': [],
 'global_phase': 0.0,
 **'memory_slots': 0,**
 'metadata': {},
 'n_qubits': 0,
 'name': 'circuit-95',
 'qreg_sizes': [],
 'qubit_labels': []}
Config:
{'memory_slots': 0, 'n_qubits': 0}

**Comment by hhorii** [comment]
Note: A qobj of the above circuit (which uses `add_bits`) was not handled in the current cloud service.

```
Traceback (most recent call last):
  Input In [23] in <cell line: 1>
    backend.run(circuit).result()
  File /opt/conda/lib/python3.8/site-packages/qiskit/providers/ibmq/job/ibmqjob.py:290 in result
    raise IBMQJobFailureError(
IBMQJobFailureError: 'Unable to retrieve result for job 637748b2c182786a7176f744. Job has failed: QObj does not conform to the schema. Error code: 1101.'

Use %tb to get the full traceback.
```

**Comment by t-imamichi** [comment]
We might need to copy this PR to terra because it seems an issue related to `assemble`.

**Comment by jakelishman** [comment]
This may be a problem in `assemble`, by I'm contending that Aer _really_ should not still be using `assemble`.  We're not going to evolve the legacy `Qobj` path to handle new dynamic circuits, and since Aer should be handling much more than the IBM hardware backends, it really doesn't make sense that it's using `Qobj` / JSON as a transport format between Python and C++.  Aer needs to evolve its C++/Python boundary at some point (ideally months ago, when we starting doing dynamic circuits), and it's just going to cause us more pain the longer we go without it.

**Comment by kevinsung** [comment]
@jakelishman Should we make a separate issue to track updating Aer's C++/Python boundary? Is that a significant undertaking? Would you be able to break it down into tasks? Is this issue completely blocked by that?

**Comment by hhorii** [comment]
#1717 resolved this issue.
